src/main/java/org/xbill/DNS/lookup/LookupSession.java

```diff
@@ -0,0 +1,299 @@
+// SPDX-License-Identifier: BSD-2-Clause
+package org.xbill.DNS.lookup;
+
+import static java.lang.String.format;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singletonList;
+import static java.util.concurrent.CompletableFuture.*;
+import static java.util.stream.Collectors.toList;
+import static org.xbill.DNS.Type.CNAME;
+import static org.xbill.DNS.Type.DNAME;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import lombok.Builder;
+import lombok.NonNull;
+import lombok.Singular;
+import org.xbill.DNS.CNAMERecord;
+import org.xbill.DNS.Cache;
+import org.xbill.DNS.Credibility;
+import org.xbill.DNS.DClass;
+import org.xbill.DNS.DNAMERecord;
+import org.xbill.DNS.Message;
+import org.xbill.DNS.Name;
+import org.xbill.DNS.NameTooLongException;
+import org.xbill.DNS.Rcode;
+import org.xbill.DNS.Record;
+import org.xbill.DNS.Resolver;
+import org.xbill.DNS.Section;
+import org.xbill.DNS.SetResponse;
+import org.xbill.DNS.Type;
+
+/**
+ * DNSSession provides facilities to make DNS Queries. A DNSSession is intended to be long lived,
+ * and it's behaviour can be modified by calling the setter methods.
+ */
+@Builder(buildMethodName = "innerBuild")
+public class LookupSession {
+  private static final int DEFAULT_MAX_ITERATIONS = 16;
+  private static final int DEFAULT_NDOTS = 1;
+  /** The {@link Resolver} to use to look up records. */
+  @NonNull private final Resolver resolver;
+  /**
+   * Sets the maximum number of CNAME or DNAME redirects allowed before lookups with fail with
+   * {@link RedirectOverflowException}.
+   */
+  @Builder.Default private final int maxRedirects = DEFAULT_MAX_ITERATIONS;
+  /**
+   * Sets the threshold for the number of dots which much appear in a name before it is considered
+   * absolute. The default is {@code 1}, meaning meaning that if there are any dots in a name, the
+   * name will be tried first as an absolute name.
+   */
+  @Builder.Default private final int ndots = DEFAULT_NDOTS;
+
+  /** Configures the search path used to look up relative names with less than ndots dots. */
+  @Singular("searchPathElement")
+  private final List<Name> searchPath;
+
+  /**
+   * If set to true, cached results with multiple records will be returned with the starting point
+   * shifted one step per request.
+   */
+  @Builder.Default private final boolean cycleResults = false;
+
+  /**
+   * Configures the Cache instances to be used for lookups for the different {@link DClass} values.
+   */
+  @Singular("cache")
+  private final Map<Integer, Cache> caches;
```

Comment:
Is there a default Cache as before? How can I set one without bothering about what a DClass is?

Reply:
This is not yet implemented. I'm thinking that there should be a static method that would give you a _default_ LookupSession with sensible defaults set, but with ~100 comments to this PR, I would prefer to have that land a bit later.